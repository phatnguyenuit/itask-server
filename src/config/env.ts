import fs from 'fs';
import path from 'path';
import dotEnv from 'dotenv';
import dotEnvExpand from 'dotenv-expand';

import logger from 'utils/logger';

const NODE_ENV = process.env.NODE_ENV;

const getFilePath = (file: string) =>
  path.resolve(String(process.env.PWD), file);

const dotEnvFiles = [
  getFilePath(`.env.${NODE_ENV}.local`),
  // Don't include `.env.local` for `test` environment
  // since normally you expect tests to produce the same
  // results for everyone
  NODE_ENV !== 'test' && getFilePath('.env.local'),
  getFilePath(`.env.${NODE_ENV}`),
  getFilePath('.env'),
].filter(Boolean) as string[];

logger.log('Loading dotEnvFiles...', dotEnvFiles);

export const loadEnv = () => {
  dotEnvFiles.forEach((dotEnvFile) => {
    if (fs.existsSync(dotEnvFile)) {
      dotEnvExpand(
        dotEnv.config({
          path: dotEnvFile,
        }),
      );
    }
  });
};

export const getEnv = (name: string, defaultValue?: string) => {
  const value = process.env[name] || defaultValue;

  if (!value) {
    throw new Error(`Environment variable named "${name}" is not defined.`);
  }

  return value;
};
