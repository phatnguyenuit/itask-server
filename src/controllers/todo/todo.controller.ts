import { Prisma } from '@prisma/client';
import { RequestHandler, Request } from 'express';

import { DEFAULT_PAGE, DEFAULT_PAGE_SIZE } from 'constants/common';
import APIError from 'constants/APIError';
import { RECORD_NOT_FOUND } from 'constants/errors';
import { getPagination } from 'utils/paginate';
import prisma from 'utils/prisma';

import { validate } from './todo.controller.types.validator';
import { validateRawData } from 'utils/common';

const queryParamsMapping = (queryParams: Request['query']) => {
  const { userId, id, isCompleted, title, page, pageSize } = queryParams;

  return {
    userId: Boolean(userId) ? Number(userId) : undefined,
    id: Boolean(id) ? Number(id) : undefined,
    title: Boolean(title) ? String(title) : undefined,
    isCompleted: Boolean(isCompleted) ? isCompleted === 'true' : undefined,
    page: Boolean(page) ? Number(page) : undefined,
    pageSize: Boolean(pageSize) ? Number(pageSize) : undefined,
  };
};

export const searchTodos: RequestHandler = async (req, res, next) => {
  try {
    const {
      userId,
      id,
      isCompleted,
      title,
      page = DEFAULT_PAGE,
      pageSize = DEFAULT_PAGE_SIZE,
    } = queryParamsMapping(req.query);

    const whereParams: Prisma.TodoWhereInput = {
      userId,
      id,
      isCompleted,
      title: {
        contains: title,
        mode: 'insensitive',
      },
    };

    const count = await prisma.todo.count({
      where: whereParams,
    });

    if (count === 0) {
      return res.json({
        data: {
          page,
          pageSize,
          total: count,
          totalPages: 0,
          data: [],
        },
      });
    }

    const { skip, take, totalPages } = getPagination(count, page, pageSize);
    const todos = await prisma.todo.findMany({
      where: whereParams,
      skip,
      take,
    });

    return res.json({
      data: {
        total: count,
        totalPages,
        page,
        pageSize,
        data: todos,
      },
    });
  } catch (error) {
    next(error);
  }
};

export const createTodo: RequestHandler = async (req, res, next) => {
  try {
    console.log("validate('CreateTodoInput')", validate('CreateTodoInput'));
    const data = validateRawData(validate('CreateTodoInput'), req.body);

    console.log(`data`, data);

    const existingUser = await prisma.user.findUnique({
      where: {
        id: data.userId,
      },
    });

    if (!existingUser) {
      throw new APIError('User is not found', 404);
    }

    const todo = await prisma.todo.create({
      data,
    });

    return res.json(todo);
  } catch (error) {
    next(error);
  }
};

type TodoIdPathParams = {
  id: string;
};

export const verifyTodoId: RequestHandler<TodoIdPathParams> = async (
  req,
  res,
  next,
) => {
  const id = Number(req.params.id);

  if (!id) {
    return next(new APIError('Wrong ID provided.', 400));
  }

  const existingTodo = await prisma.todo.findUnique({
    where: { id },
  });

  if (!existingTodo) {
    return next(RECORD_NOT_FOUND);
  }

  next();
};

export const updateTodo: RequestHandler<TodoIdPathParams> = async (
  req,
  res,
  next,
) => {
  try {
    const id = Number(req.params.id);

    const { title, isCompleted } = req.body;

    const todo = await prisma.todo.update({
      data: {
        title,
        isCompleted,
      },
      where: { id },
    });

    return res.json(todo);
  } catch (error) {
    next(error);
  }
};

export const getTodo: RequestHandler<TodoIdPathParams> = async (
  req,
  res,
  next,
) => {
  try {
    const id = Number(req.params.id);
    const existingTodo = await prisma.todo.findUnique({
      where: { id },
    });

    return res.json(existingTodo);
  } catch (error) {
    next(error);
  }
};

export const deleteTodo: RequestHandler<TodoIdPathParams> = async (
  req,
  res,
  next,
) => {
  try {
    const id = Number(req.params.id);

    await prisma.todo.delete({
      where: { id },
    });

    return res.json({
      message: 'Request successfully.',
    });
  } catch (error) {
    next(error);
  }
};
