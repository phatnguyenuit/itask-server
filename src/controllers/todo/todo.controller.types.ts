export type CreateTodoInput = {
  userId: number;
  title: string;
  isCompleted: boolean;
};
