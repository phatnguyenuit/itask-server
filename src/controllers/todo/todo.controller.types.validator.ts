/* tslint:disable */
// generated by typescript-json-validator
import Ajv from 'ajv';
import { CreateTodoInput } from './todo.controller.types';
export const ajv = new Ajv({
  allErrors: true,
  coerceTypes: false,
  format: 'fast',
  nullable: true,
  unicode: true,
  uniqueItems: true,
  useDefaults: true,
});

ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));

export { CreateTodoInput };
export const Schema = {
  $id: './src/controllers/todo/todo.controller.types.ts',
  $schema: 'http://json-schema.org/draft-07/schema#',
  definitions: {
    CreateTodoInput: {
      additionalProperties: false,
      defaultProperties: [],
      properties: {
        isCompleted: {
          type: 'boolean',
        },
        title: {
          type: 'string',
        },
        userId: {
          type: 'number',
        },
      },
      required: ['isCompleted', 'title', 'userId'],
      type: 'object',
    },
  },
};
ajv.addSchema(Schema, 'Schema');
export function validate(
  typeName: 'CreateTodoInput',
): (value: unknown) => CreateTodoInput;
export function validate(typeName: string): (value: unknown) => any {
  const validator: any = ajv.getSchema(`Schema#/definitions/${typeName}`);
  return (value: unknown): any => {
    if (!validator) {
      throw new Error(
        `No validator defined for Schema#/definitions/${typeName}`,
      );
    }

    const valid = validator(value);

    if (!valid) {
      throw new Error(
        'Invalid ' +
          typeName +
          ': ' +
          ajv.errorsText(
            validator.errors!.filter((e: any) => e.keyword !== 'if'),
            { dataVar: typeName },
          ),
      );
    }

    return value as any;
  };
}
