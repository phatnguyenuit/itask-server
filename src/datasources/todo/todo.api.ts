import { getEnv } from 'config/env';
import { CreateTodoPayload, SearchTodoParams, UpdateTodoPayload } from 'shared';
import { NonNullableDeep } from '../../typings/common';

import BaseAPI from '../base.api';
import { validate } from './todo.api.types.validator';

class TodoAPI extends BaseAPI {
  constructor() {
    super();
    this.baseURL = `${getEnv('ITASK_API_URL')}/api/v1/todos`;
  }

  async searchTodos(
    userId: number,
    params: Partial<NonNullableDeep<SearchTodoParams>>,
  ) {
    const response = this.logOrThrowValidationError(
      validate('SearchTodosResponse'),
      await this.get('/', { ...params, userId }),
    );

    return response.data;
  }

  async createTodo(payload: CreateTodoPayload) {
    const response = this.logOrThrowValidationError(
      validate('CreateTodoResponse'),
      await this.post('/', payload),
    );

    return response;
  }

  async getTodo(id: number) {
    const response = this.logOrThrowValidationError(
      validate('GetTodoResponse'),
      await this.get(`/${id}`),
    );

    return response;
  }

  async updateTodo(id: number, payload: UpdateTodoPayload) {
    const response = this.logOrThrowValidationError(
      validate('UpdateTodoResponse'),
      await this.patch(`/${id}`, payload),
    );

    return response;
  }

  async deleteTodo(id: number) {
    await this.delete(`/${id}`);

    return true;
  }
}

export default TodoAPI;
