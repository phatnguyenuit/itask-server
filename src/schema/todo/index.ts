import { resolvers } from './todo.resolvers';
import typeDefs from './todo.typeDefs.gql';

export { resolvers, typeDefs };
