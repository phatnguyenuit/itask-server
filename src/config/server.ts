import { ApolloServerExpressConfig } from 'apollo-server-express';
import {
  ApolloServerPluginLandingPageDisabled,
  ApolloServerPluginDrainHttpServer,
  PluginDefinition,
} from 'apollo-server-core';
import http from 'http';

import app from '../app';
import dataSources from '../datasources';
import schema from '../schema';
// import prisma from './utils/prisma';

export const httpServer = http.createServer(app);

export const getApolloPlugins = () =>
  [
    ApolloServerPluginDrainHttpServer({ httpServer }),
    process.env.NODE_ENV === 'production'
      ? ApolloServerPluginLandingPageDisabled // Disable landing page for production mode
      : undefined,
  ].filter(Boolean) as PluginDefinition[];

export const apolloConfig: ApolloServerExpressConfig = {
  schema,
  dataSources: () => dataSources,
  plugins: getApolloPlugins(),
  nodeEnv: process.env.NODE_ENV,
  context: ({ req }) => ({
    headers: req.headers,
    // prisma,  // Prisma client context
  }),
};
