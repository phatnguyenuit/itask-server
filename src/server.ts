import { ApolloServer, ApolloServerExpressConfig } from 'apollo-server-express';
import {
  ApolloServerPluginLandingPageDisabled,
  ApolloServerPluginDrainHttpServer,
  PluginDefinition,
} from 'apollo-server-core';
import express from 'express';
import http from 'http';

import app from './app';
import { loadEnv, getEnv } from './config/env';
import dataSources from './datasources';
import schema from './schema';
// import prisma from './utils/prisma';

// Ensure environment variables are read.
loadEnv();

const httpServer = http.createServer(app);

const apolloPlugins = [
  ApolloServerPluginDrainHttpServer({ httpServer }),
  process.env.NODE_ENV === 'production'
    ? ApolloServerPluginLandingPageDisabled // Disable landing page for production mode
    : undefined,
].filter(Boolean) as PluginDefinition[];

const apolloConfig: ApolloServerExpressConfig = {
  schema,
  dataSources: () => dataSources,
  plugins: apolloPlugins,
  nodeEnv: process.env.NODE_ENV,
  context: ({ req }) => ({
    headers: req.headers,
    // prisma, // Prisma client context
  }),
};

async function startApolloServer(
  config: ApolloServerExpressConfig,
  app: express.Application,
) {
  const server = new ApolloServer(config);

  await server.start();
  server.applyMiddleware({ app });

  const PORT = Number(getEnv('PORT', '4000'));
  await new Promise<void>((resolve) =>
    httpServer.listen({ port: PORT }, resolve),
  );

  console.log(
    `ðŸš€ Server ready at http://localhost:${PORT}${server.graphqlPath}`,
  );
}

startApolloServer(apolloConfig, app);
