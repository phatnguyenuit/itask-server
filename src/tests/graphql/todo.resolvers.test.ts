import { ApolloServer } from 'apollo-server-express';
import { gql } from 'graphql-tag';

import { apolloConfig } from 'config/server';
import { MOCK_EXPRESS_CONTEXT } from 'fixtures/context';
import { MOCK_TODO, MOCK_TODOS } from 'fixtures/todo';
import {
  createCreateTodoSuccessHandler,
  createGetTodoSuccessHandler,
  createSearchTodosSuccessHandler,
  createUpdateTodoSuccessHandler,
  deleteTodoSuccessHandler,
} from 'mocks/handlers/todo.handlers';
import { mockServer } from 'mocks/server';

describe('resolvers/todo', () => {
  let apolloServer: ApolloServer;

  beforeAll(() => {
    apolloServer = new ApolloServer(apolloConfig);
  });

  describe('searchTodos', () => {
    const SEARCH_TODOS = gql`
      query searchTodos($userId: Int!, $searchParams: SearchTodoParams) {
        searchTodos(userId: $userId, searchParams: $searchParams) {
          page
          pageSize
          total
          totalPages
          todos {
            id
            userId
            title
            isCompleted
          }
        }
      }
    `;

    it('should search todos successfully', async () => {
      mockServer.use(
        createSearchTodosSuccessHandler({
          total: MOCK_TODOS.length,
          totalPages: 1,
          page: 1,
          pageSize: 10,
          data: MOCK_TODOS,
        }),
      );
      // run query against the server and snapshot the output
      const res = await apolloServer.executeOperation(
        {
          query: SEARCH_TODOS,
          variables: { userId: 1 },
        },
        MOCK_EXPRESS_CONTEXT,
      );

      expect(res).toMatchSnapshot();
    });
  });

  describe('getTodo', () => {
    const GET_TODO = gql`
      query getTodo($getTodoId: Int!) {
        getTodo(id: $getTodoId) {
          id
          userId
          title
          isCompleted
        }
      }
    `;

    it('should get todo by id successfully', async () => {
      mockServer.use(createGetTodoSuccessHandler(MOCK_TODO));
      // run query against the server and snapshot the output
      const res = await apolloServer.executeOperation(
        {
          query: GET_TODO,
          variables: { getTodoId: 1 },
        },
        MOCK_EXPRESS_CONTEXT,
      );

      expect(res).toMatchSnapshot();
    });
  });

  describe('createTodo', () => {
    const CREATE_TODO = gql`
      mutation createTodo($payload: CreateTodoPayload!) {
        createTodo(payload: $payload) {
          id
          userId
          title
          isCompleted
        }
      }
    `;

    it('should create todo successfully', async () => {
      mockServer.use(createCreateTodoSuccessHandler(MOCK_TODO));
      // run query against the server and snapshot the output
      const res = await apolloServer.executeOperation(
        {
          query: CREATE_TODO,
          variables: {
            payload: {
              userId: 1,
              title: 'todo 1',
              isCompleted: false,
            },
          },
        },
        MOCK_EXPRESS_CONTEXT,
      );

      expect(res).toMatchSnapshot();
    });
  });

  describe('updateTodo', () => {
    const UPDATE_TODO = gql`
      mutation updateTodo($updateTodoId: Int!, $payload: UpdateTodoPayload!) {
        updateTodo(id: $updateTodoId, payload: $payload) {
          id
          userId
          title
          isCompleted
        }
      }
    `;

    it('should update todo successfully', async () => {
      mockServer.use(createUpdateTodoSuccessHandler(MOCK_TODO));
      // run query against the server and snapshot the output
      const res = await apolloServer.executeOperation(
        {
          query: UPDATE_TODO,
          variables: {
            updateTodoId: 1,
            payload: {
              title: 'todo 1',
            },
          },
        },
        MOCK_EXPRESS_CONTEXT,
      );

      expect(res).toMatchSnapshot();
    });
  });

  describe('deleteTodo', () => {
    const DELETE_TODO = gql`
      mutation deleteTodo($deleteTodoId: Int!) {
        deleteTodo(id: $deleteTodoId)
      }
    `;

    it('should delete todo successfully', async () => {
      mockServer.use(deleteTodoSuccessHandler);
      // run query against the server and snapshot the output
      const res = await apolloServer.executeOperation(
        {
          query: DELETE_TODO,
          variables: {
            deleteTodoId: 1,
          },
        },
        MOCK_EXPRESS_CONTEXT,
      );

      expect(res).toMatchSnapshot();
    });
  });
});
