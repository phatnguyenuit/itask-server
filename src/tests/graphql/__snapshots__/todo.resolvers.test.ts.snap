// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvers/todo createTodo should create todo successfully 1`] = `
Object {
  "data": Object {
    "createTodo": Object {
      "id": 1,
      "isCompleted": false,
      "title": "todo 1",
      "userId": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolvers/todo deleteTodo should delete todo successfully 1`] = `
Object {
  "data": Object {
    "deleteTodo": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolvers/todo getTodo should get todo by id successfully 1`] = `
Object {
  "data": Object {
    "getTodo": Object {
      "id": 1,
      "isCompleted": false,
      "title": "todo 1",
      "userId": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolvers/todo searchTodos should search todos failed 1`] = `
Object {
  "data": Object {
    "searchTodos": null,
  },
  "errors": Array [
    [GraphQLError: 500: Internal Server Error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolvers/todo searchTodos should search todos successfully 1`] = `
Object {
  "data": Object {
    "searchTodos": Object {
      "page": 1,
      "pageSize": 10,
      "todos": Array [
        Object {
          "id": 1,
          "isCompleted": false,
          "title": "todo 1",
          "userId": 1,
        },
        Object {
          "id": 2,
          "isCompleted": false,
          "title": "todo 2",
          "userId": 1,
        },
      ],
      "total": 2,
      "totalPages": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`resolvers/todo updateTodo should update todo successfully 1`] = `
Object {
  "data": Object {
    "updateTodo": Object {
      "id": 1,
      "isCompleted": false,
      "title": "todo 1",
      "userId": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
